src/app/content/practiceQuestions/utils.ts

```diff
@@ -0,0 +1,54 @@
+import { isLinkedArchiveTreeSection } from '../guards';
+import { Book, LinkedArchiveTreeSection } from '../types';
+import { findArchiveTreeNodeById, flattenArchiveTree } from '../utils/archiveTreeUtils';
+import { stripIdVersion } from '../utils/idUtils';
+import { PracticeQuestionsSummary } from './types';
+
+export const pageHasPracticeQuestions = (pageId: string, summary: PracticeQuestionsSummary) => {
+  return Boolean(summary.countsPerSource[pageId]);
+};
+
+export const getPracticeQuestionsLocationFilters = (
+  summary: PracticeQuestionsSummary | null, book: Book | undefined
+) => {
+  // key is an id of a parent of sections stored in a value
+  const locationFilters: Map<string, LinkedArchiveTreeSection[]> = new Map();
+
+  if (!book || !summary) { return locationFilters; }
+
+  const tree = flattenArchiveTree(book.tree);
+
+  for (const node of tree) {
+    if (isLinkedArchiveTreeSection(node) && summary.countsPerSource[stripIdVersion(node.id)]) {
+      const parentId = stripIdVersion(node.parent.id);
+      locationFilters.set(parentId, [...locationFilters.get(parentId) || [], node]);
+    }
+  }
+
+  return locationFilters;
+};
+
+const flattenLocationFilters = (locationFilters: Map<string, LinkedArchiveTreeSection[]>) => {
+  let flattened: LinkedArchiveTreeSection[] = [];
+  for (const sections of locationFilters.values()) {
+    flattened = flattened.concat(sections);
+  }
+  return flattened;
+};
+
+export const getNextPageWithPracticeQuestions = (
+  nodeId: string, locationFilters: Map<string, LinkedArchiveTreeSection[]>, book: Book | undefined
+) => {
+  if (!book) { return; }
+
+  const node = findArchiveTreeNodeById(book.tree, nodeId);
+
+  if (!node) { return; }
+
+  const flatLocationFilters = flattenLocationFilters(locationFilters);
```

Comment:
doesn't this contain tree sections? can't you look for the index + 1 in here instead of re-flattening the tree and cross referencing them?

Reply:
It contain only tree sections that have practice questions so if `node` is a section like `Preface` we will not be able to find id it here

This is why I search for `nodeIndex` in a `tree` and then iterate over all sections after this one and check if any of them is in `locationFilters`